### DISCUSSION TITLE
What change triggers a redeploy of a Kubernetes Configmap?

### DISCUSSION THREAD

**Discussion by sheldonhull** [pulumi/pulumi#8952] [Pulumi Providers] [answered]
What triggers a configmap to actually recreate? Any articles on this?

I'm creating a config.yml file via the ConfigMap populated from the Pulumi configuration.
If I update a value in the pulumi config it reruns the plan, but doesn't seem to regenerate the new ConfigMap on demand.

1. What triggers a change detection for a ConfigMap? (Link to source is fine)
2. What should I do to ensure ConfigMap is regenerated when I update the appropriate pulumi configuration values?

**Comment by rawkode** [comment, marked as answer]
This was discussed on the Pulumi Community Slack, but I'll answer this here so that others can learn from this discussion too.

ConfigMap's with Pulumi are, at the moment, **considered immutable by default when you provide** `metadata.name` (Even when the immutable flag in the spec isn't true).

This means that when you change the data, the ConfigMap will be recreated.

If you allow Pulumi to auto-name the ConfigMap, the ConfigMap will be updated in-place.

Related issues:
- https://github.com/pulumi/pulumi-kubernetes/issues/1775
- https://github.com/pulumi/pulumi-kubernetes/issues/1568#issuecomment-932424463
